File: typescript/src/ships/shield-capacitor.ts
/**
 * **The SYS capacitor** — what pips to SYS buy a raised shield, and what the shield is
 * worth once they are folded in.
 *
 * SYS does two things at once. It **resists**: the allocation adds a flat resistance to
 * every damage type, which multiplies with the shield's own stack rather than adding to
 * it, so incoming damage is scaled by `(1 − shieldResistance) × (1 − sysResistance)`. And
 * it **recharges**: the capacitor's energy is what the generator spends putting the
 * shield back up, on the same non-linear pip curve the other two capacitors follow,
 * `ratedRecharge × (pips / 4) ^ 1.1`.
 *
 * ```text
 * sysResistance         = 0.6 × (pips / 4) ^ 0.85
 * effectiveResistance   = 1 − (1 − shieldResistance) × (1 − sysResistance)
 * effectiveHitPoints    = strength / (1 − effectiveResistance)
 * ```
 *
 * The bare shield those effective figures are built from is {@link shieldMetrics}, which
 * is pip-free. What the recharge reported here *buys* — seconds from collapse back to a
 * raised shield — is {@link shieldRecovery}, which takes the same allocation and the
 * generator's regeneration rates.
 *
 * This module is data-free. {@link BuildMetrics.shieldCapacitorMetrics} (in
 * `./build-metrics`) resolves the powered generator, boosters and distributor for you.
 *
 * @remarks
 * Reference implementation: EDCD/Coriolis, `src/app/shipyard/Calculations.js`
 * (`sysResistance`, `calcShieldStrength`), commit
 * `68c042ca6e3db62372cbbb2077cf972345511712`. The algorithm is ported as fact, not code;
 * credit and licence terms are in
 * [ATTRIBUTIONS.md](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/ATTRIBUTIONS.md).
 *
 * @packageDocumentation
 */

import { capacitorRechargeAtPips } from './internal/capacitor-recharge.js';
import { requirePips } from './internal/pips.js';
import {
    effectiveHitPoints,
    mapDamageTypes,
    systemsResistance,
    type DamageResistances,
    type DamageType,
    type DamageTypeValues,
} from './resistances.js';

/** Everything {@link shieldCapacitorMetrics} needs about one raised shield. */
export interface ShieldCapacitorInput {
    /**
     * The shield's total strength, in megajoules — finite and non-negative.
     * {@link ShieldMetrics.strength}.
     */
    readonly strength: number;
    /**
     * The shield's own stacked resistances, pip-free — every one a finite fraction.
     * {@link ShieldMetrics.resistances}.
     */
    readonly resistances: DamageResistances;
    /** SYS-capacitor capacity, in megajoules — finite and non-negative. */
    readonly systemsCapacity: number;
    /** Four-SYS-pip recharge rate, in megajoules per second — finite and non-negative. */
    readonly systemsRecharge: number;
    /** Pips assigned to SYS, in `[0, 4]`. Fractional pips are accepted; defaults to `4`. */
    readonly systemsPips?: number;
}

/**
 * What one SYS allocation is worth to a raised shield.
 *
 * @remarks
 * Frozen — nested records and lists included — so a result can be held, cached and
 * shared without a defensive copy. Derive a changed figure with a spread rather than
 * by assigning into one.
 */
export interface ShieldCapacitorMetrics {
    /** Pips assigned to SYS for this result, in `[0, 4]`. */
    readonly systemsPips: number;
    /** SYS-capacitor capacity, in megajoules. */
    readonly capacity: number;
    /** Actual recharge rate at {@link systemsPips}, in megajoules per second. */
    readonly rechargeRate: number;
    /**
     * The resistance the pips contribute on their own, as a fraction: `0` at no pips
     * rising to `0.6` at four. Unrounded, like every resistance here.
     */
    readonly systemsResistance: number;
    /**
     * The shield's resistances with the pips folded in — the figures the game's own
     * panel shows while the allocation stands.
     *
     * @remarks
     * Fractions rather than percentages, and **unrounded**. The pips multiply with the
     * shield's stack rather than adding to it, so these are
     * `1 − (1 − shieldResistance) × (1 − systemsResistance)` and not a sum.
     */
    readonly effectiveResistances: DamageResistances;
    /**
     * Effective hit points against each damage type, in megajoules, behind
     * {@link ShieldCapacitorMetrics.effectiveResistances} — the raw damage of that type
     * the shield soaks at this allocation. `Infinity` where a resistance reaches 100%.
     */
    readonly effectiveHitPoints: DamageTypeValues;
}

const DAMAGE_TYPES: readonly DamageType[] = ['kinetic', 'thermal', 'explosive', 'caustic'];

function requireFiniteNonNegative(scope: string, name: string, value: number): void {
    if (!Number.isFinite(value) || value < 0) {
        throw new RangeError(`${scope}: ${name} must be a finite non-negative number`);
    }
}

/**
 * Calculate what a SYS-pip allocation gives a raised shield.
 *
 * @param input - The bare shield's strength and stacked resistances, the distributor's
 * SYS capacity and rated four-pip recharge, and the allocation to model.
 * @returns The {@link ShieldCapacitorMetrics}. At **no** pips `systemsResistance` and
 * `rechargeRate` are `0` and the effective figures equal the bare ones that went in.
 * @throws {RangeError} If `systemsPips` is outside `[0, 4]`, if a strength, capacity or
 * recharge is negative or non-finite, or if a resistance is not a finite number. The
 * pips are checked first, so a bad allocation is reported before the shield.
 * @throws {TypeError} If `resistances` is not an object carrying the four damage types.
 * @example
 * ```ts
 * import { shieldMetrics } from '@elite-dangerous-almanac/core/ships/shields';
 * import { shieldCapacitorMetrics } from '@elite-dangerous-almanac/core/ships/shield-capacitor';
 *
 * // The bare shield, then what four pips to SYS make of it.
 * const shields = shieldMetrics({
 *   hullMass: 400,
 *   baseShieldStrength: 350,
 *   generator: {
 *     minMass: 270, optMass: 540, maxMass: 1350,
 *     minMultiplier: 0.7, optMultiplier: 1.2, maxMultiplier: 1.7,
 *     kineticResistance: 0.4,
 *   },
 * });
 * const sys = shieldCapacitorMetrics({
 *   ...shields,
 *   systemsCapacity: 41,
 *   systemsRecharge: 3.9,
 *   systemsPips: 4,
 * });
 * Number(sys.systemsResistance.toFixed(2));            // -> 0.6
 * Number(sys.effectiveResistances.kinetic.toFixed(2)); // -> 0.76
 * sys.effectiveHitPoints.kinetic > shields.effectiveHitPoints.kinetic; // -> true
 * ```
 */
export function shieldCapacitorMetrics(input: ShieldCapacitorInput): ShieldCapacitorMetrics {
    const scope = 'shieldCapacitorMetrics';
    const systemsPips = input.systemsPips ?? 4;
    // Named for the parameter the caller wrote, and checked before the shield so a
    // build with no shields still reports a bad allocation.
    requirePips(scope, 'systemsPips', systemsPips);
    for (const field of ['strength', 'systemsCapacity', 'systemsRecharge'] as const) {
        requireFiniteNonNegative(scope, field, input[field]);
    }
    const resistances: unknown = input.resistances;
    if (typeof resistances !== 'object' || resistances === null) {
        throw new TypeError(`${scope}: resistances must carry the four damage types`);
    }
    for (const type of DAMAGE_TYPES) {
        if (!Number.isFinite(input.resistances[type])) {
            throw new RangeError(`${scope}: resistances.${type} must be a finite number`);
        }
    }

    const sysResistance = systemsResistance(systemsPips);
    // The SYS pips multiply with the stacked shield resistance rather than adding to it.
    const effectiveResistances: DamageResistances = Object.freeze(
        mapDamageTypes((type) => 1 - (1 - input.resistances[type]) * (1 - sysResistance)),
    );
    return Object.freeze({
        systemsPips,
        capacity: input.systemsCapacity,
        rechargeRate: capacitorRechargeAtPips(input.systemsRecharge, systemsPips),
        systemsResistance: sysResistance,
        effectiveResistances,
        effectiveHitPoints: Object.freeze(effectiveHitPoints(input.strength, effectiveResistances)),
    });
}

File: dotnet/src/EliteDangerousAlmanac/Ships/ShieldCapacitor.cs
using System;
using EliteDangerousAlmanac.Ships.Internal;

namespace EliteDangerousAlmanac.Ships;

/// <summary>Everything a SYS-capacitor calculation needs about one raised shield.</summary>
/// <param name="Strength">The shield's total strength, in megajoules.</param>
/// <param name="Resistances">The shield's own stacked resistances, pip-free, each a finite fraction.</param>
/// <param name="SystemsCapacity">The SYS-capacitor capacity, in megajoules.</param>
/// <param name="SystemsRecharge">The four-SYS-pip recharge rate, in megajoules per second.</param>
public sealed record ShieldCapacitorInput(
    double Strength,
    DamageTypeValues Resistances,
    double SystemsCapacity,
    double SystemsRecharge)
{
    /// <summary>The bare shield a build already holds, with the distributor's SYS figures.</summary>
    /// <param name="shields">The pip-free shield metrics.</param>
    /// <param name="systemsCapacity">The SYS-capacitor capacity, in megajoules.</param>
    /// <param name="systemsRecharge">The four-SYS-pip recharge rate, in megajoules per second.</param>
    /// <returns>The input, so the two halves compose without a hand-written adapter.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="shields"/> is <see langword="null"/>.</exception>
    public static ShieldCapacitorInput For(
        ShieldMetrics shields, double systemsCapacity, double systemsRecharge)
    {
        if (shields is null) throw new ArgumentNullException(nameof(shields));

        return new ShieldCapacitorInput(
            shields.Strength, shields.Resistances, systemsCapacity, systemsRecharge);
    }
}

/// <summary>What one SYS allocation is worth to a raised shield.</summary>
/// <param name="SystemsPips">The pips assigned to SYS for this result, from zero through four.</param>
/// <param name="Capacity">The SYS-capacitor capacity, in megajoules.</param>
/// <param name="RechargeRate">The recharge rate at the allocation, in megajoules per second.</param>
/// <param name="SystemsResistance">
/// The resistance the pips contribute on their own, as a fraction: none at no pips, rising to
/// three fifths at four.
/// </param>
/// <param name="EffectiveResistances">
/// The shield's resistances with the pips folded in, which are the figures the game's own panel
/// shows while the allocation stands. The pips multiply with the shield's stack rather than adding
/// to it, so these are not a sum.
/// </param>
/// <param name="EffectiveHitPoints">
/// The raw damage of each type the shield soaks at this allocation, in megajoules. It is infinite
/// where a resistance reaches the whole of that damage type.
/// </param>
/// <remarks>Every resistance here is a fraction rather than a percentage, and unrounded.</remarks>
public sealed record ShieldCapacitorMetrics(
    double SystemsPips,
    double Capacity,
    double RechargeRate,
    double SystemsResistance,
    DamageTypeValues EffectiveResistances,
    DamageTypeValues EffectiveHitPoints);

/// <summary>
/// The SYS capacitor: what pips to SYS buy a raised shield, and what the shield is worth once they
/// are folded in.
/// </summary>
/// <remarks>
/// <para>
/// SYS does two things at once. It resists: the allocation adds a flat resistance to every damage
/// type, which multiplies with the shield's own stack rather than adding to it. And it recharges:
/// the capacitor's energy is what the generator spends putting the shield back up, on the same
/// non-linear pip curve the other two capacitors follow.
/// </para>
/// <para>
/// The bare shield those effective figures are built from is <see cref="Shields.Metrics"/>, which
/// is pip-free. What the recharge reported here buys, in seconds from collapse back to a raised
/// shield, is <see cref="ShieldRecovery.Metrics"/>, which takes the same allocation and the
/// generator's regeneration rates.
/// </para>
/// <para>
/// Reference implementation: EDCD/Coriolis, <c>src/app/shipyard/Calculations.js</c>. The algorithm
/// is ported as fact rather than as code. See <c>ATTRIBUTIONS.md</c> for credit and licence terms.
/// </para>
/// </remarks>
public static class ShieldCapacitor
{
    /// <summary>Calculates what a SYS-pip allocation gives a raised shield.</summary>
    /// <param name="input">
    /// The bare shield's strength and stacked resistances, and the distributor's SYS capacity and
    /// rated four-pip recharge.
    /// </param>
    /// <param name="systemsPips">
    /// The pips assigned to SYS, from zero through four. Fractional pips are accepted.
    /// </param>
    /// <returns>
    /// What the allocation is worth. At no pips the resistance and the recharge rate are zero, and
    /// the effective figures equal the bare ones that went in.
    /// </returns>
    /// <remarks>
    /// The pips are checked before the shield, so a bad allocation is reported first.
    /// </remarks>
    /// <exception cref="ArgumentNullException"><paramref name="input"/> is <see langword="null"/>.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The allocation is outside zero through four, a strength, capacity or recharge rate is
    /// negative or not finite, or a resistance is not a finite number.
    /// </exception>
    public static ShieldCapacitorMetrics Metrics(ShieldCapacitorInput input, double systemsPips = 4)
    {
        // Checked before the shield, so a build with none still reports a bad allocation.
        RangeGuards.RequirePips(systemsPips, nameof(systemsPips));
        if (input is null) throw new ArgumentNullException(nameof(input));

        RangeGuards.RequireFiniteNonNegative(input.Strength, "input.Strength");
        RangeGuards.RequireFiniteNonNegative(input.SystemsCapacity, "input.SystemsCapacity");
        RangeGuards.RequireFiniteNonNegative(input.SystemsRecharge, "input.SystemsRecharge");
        if (input.Resistances is null) throw new ArgumentNullException("input.Resistances");
        foreach (DamageType type in new[]
                 {
                     DamageType.Kinetic, DamageType.Thermal, DamageType.Explosive, DamageType.Caustic,
                 })
        {
            double resistance = input.Resistances[type];
            if (double.IsNaN(resistance) || double.IsInfinity(resistance))
            {
                throw new ArgumentOutOfRangeException(
                    $"input.Resistances.{type}", resistance, "The value must be a finite number.");
            }
        }

        double systemsResistance = Resistances.SystemsResistance(systemsPips);
        // The SYS pips multiply with the stacked shield resistance rather than adding to it.
        DamageTypeValues effectiveResistances = Resistances.MapDamageTypes(
            type => 1 - ((1 - input.Resistances[type]) * (1 - systemsResistance)));

        return new ShieldCapacitorMetrics(
            SystemsPips: systemsPips,
            Capacity: input.SystemsCapacity,
            RechargeRate: CapacitorRecharge.AtPips(input.SystemsRecharge, systemsPips),
            SystemsResistance: systemsResistance,
            EffectiveResistances: effectiveResistances,
            EffectiveHitPoints: Resistances.EffectiveHitPoints(input.Strength, effectiveResistances));
    }
}
